docs/kb/laptop-dev-env.md: add NovaCustom dev env doc (#1005)
Dasharo/docs: docs/kb/laptop-dev-env.md

# NovaCustom laptop development environment

This document describes the process of configuring a development and debugging
environment for NovaCustom laptops.

> The instruction is applicable for workstations running Fedora. If you're using

Reviewer: ```suggestion
The instruction is applicable for PCs running Fedora. If you're using
```

Author: https://github.com/Dasharo/docs/compare/4e9ec5622c840b8a6d840a400efafdd5de3ee80f..6a19fdc3b9c3ed99815a29ce364e8e90543a5ac2

a different distro, some commands (e.g. package installation) may be different.

## Embedded Controller

This section describes the development environment for EC firmware on NovaCustom

> devices. The Embedded Controller firmware is heavily based on System76 EC

Reviewer: ```suggestion
devices. The Embedded Controller firmware is heavily based on [System76 EC](https://github.com/system76/ec)
```

Providing a reference seems like a right thing to do.

Author: https://github.com/Dasharo/docs/compare/4e9ec5622c840b8a6d840a400efafdd5de3ee80f..6a19fdc3b9c3ed99815a29ce364e8e90543a5ac2

firmware and makes use of open-source debugging tools and software also
developed by System76.

### Flashing

This section outlines the parts and software needed to perform external flashing
of the Embedded Controller.

| Size | 14 inch models | 16 inch models |
| --- | --- | --- |
| Programmer | Arduino Mega 2560 | Arduino Mega 2560 |

> | Cable | **FFC 0.5mm pitch 24 pin** | **FFC 1.0mm pitch 24 pin** |

Reviewer: What about the adapter from the FFC to arduino pin header?

Grounding cable?
USB-B to USB-A for Arduino?

Author: Added the rest of the cables to the table: https://github.com/Dasharo/docs/compare/4e9ec5622c840b8a6d840a400efafdd5de3ee80f..6a19fdc3b9c3ed99815a29ce364e8e90543a5ac2

Reviewer: I still don't see the adapter on the list. Currently it would be impossible to connect the FFC cable to arduino.

| Software | ecflash | ecflash |

#### Preparation

1. Obtain a copy of the EC firmware and tools:

    ```bash
    git clone https://github.com/Dasharo/ec.git
    cd ec
    git submodule update --init --checkout
    ./scripts/deps.sh
    ```

1. Follow the on-screen instructions to finish configuring your development
environment.
1. Connect the Arduino to your workstation with a USB-A to USB-B cable. Do not
connect the target laptop yet.
1. Flash the Arduino with debugger firmware:

    ```bash
    make BOARD=arduino/mega2560
    make BOARD=arduino/mega2560 flash
    ```

#### Flash the target laptop

> 1. In the EC directory, build the EC flashing application:

Reviewer: I thing this goes to preparation

Author: Right, moved: https://github.com/Dasharo/docs/compare/4e9ec5622c840b8a6d840a400efafdd5de3ee80f..6a19fdc3b9c3ed99815a29ce364e8e90543a5ac2

    ```bash
    cargo build --manifest-path ecflash/Cargo.toml --example isp --release
    ```

1. Connect the Arduino to the target laptop with the appropriate FFC cable for
   your model. If the cable is narrower than the socket on the laptop, align the
   cable and socket to pin 0 (to the left).

> 1. Connect a second cable (e.g. USB-C to USB-C) from the workstation to the

Reviewer: Is it safe to connect VBUS with host and target?

Author: with usb-c it should be safe, because vbus stays at 0V until power roles are determined

Reviewer: What happens if the direction cannot be determined?

Author: USB-PD spec says Vbus shall be vSafe0V or vSafe5V in an error condition. 5V is safe, because the PD controller will not open input or output switches until it negotiates 20V.

   target - this cable is needed provide grounding.
1. Flash the target:

    ```bash
    sudo ecflash/target/release/examples/isp [path/to/firmware.rom]
    ```

##### Troubleshooting

In case flashing fails, try the following:

- Try running the command again, sometimes it starts working on the third
  attempt
- Try another grounding cable, USB-A to USB-C, or even HDMI to HDMI have been
  shown to work
- Try with / without the internal battery connected
- Try with / without the AC adapter connected
- Try plugging in the AC adapter right before / after running the command (last
  resort)

### Building

Build EC firmware:

```bash

> make BOARD=your/board

Reviewer: I would add a list of laptops and their target BOARD names

Author: Done: https://github.com/Dasharo/docs/compare/4e9ec5622c840b8a6d840a400efafdd5de3ee80f..6a19fdc3b9c3ed99815a29ce364e8e90543a5ac2

```

### Debugging

#### Build EC firmware with debugger support enabled

1. In file `src/board/system76/common/common.mk`, enable parallel port debugging

>     and set the log level to DEBUG:

Reviewer: Only debug? I would rephrase to:

```suggestion
    and choose the log level to get the verbosity you need:
```

    ```makefile
    # Set log level
    # 0 - NONE
    # 1 - ERROR
    # 2 - WARN
    # 3 - INFO
    # 4 - DEBUG
    # 5 - TRACE
    CFLAGS+=-DLEVEL=4

    # Uncomment to enable debug logging over keyboard parallel port
    CFLAGS+=-DPARALLEL_DEBUG

    # Uncomment to enable I2C debug on 0x76
    #CFLAGS+=-DI2C_DEBUGGER=0x76
    ```

1. Build firmware for the target:

    ```bash
    make BOARD=your/board
    ```

1. Flash the firmware to the target:

    ```bash
    make BOARD=your/board flash_external
    # or
    sudo ecflash/target/release/examples/isp path/to/your/ec.rom
    ```

1. Start the console:

    ```bash
    make BOARD=your/board console_external
    ```

> 1. Power on the target and observe the console for logs.

Reviewer: Hmm AFAIK, the flashing has to be done on disconnected battery. The logs start to appear once AC is plugged.

Author: Kind of. The logs will be printed if the EC finds the debugger while powering on, so theoretically holding the power button for 10 seconds to reset the EC will work too. But you're right, this should be noted here.

Reviewer: If you update this section, let me know.

Author: 6cac4ed62ff5ffcfc40e48cf691852b91aa29dbe

## BIOS

This section describes the development environment for x86 firmware on
NovaCustom devices.

### Flashing

This section outlines the parts and software needed to perform external flashing
of the BIOS.

BIOS flashing is the same as with other platforms, except on ADL and MTL models,
where the BIOS chip has a WSON-8 form factor. The following table of specs is
provided for convenience:

| CPU generation | Tiger Lake | Alder Lake | Meteor Lake |
| --- | --- | --- | --- |
| Programmer | CH341a | CH341a | **CH341a v1.2 with voltage switch** |
| Probe | SOIC-8 | **WSON-8** | **WSON-8** |
| Voltage | 3.3v | 3.3v | **1.8v** |
| Software | flashrom | flashrom | flashrom |

Use flashrom with the CH341a programmer:

```bash
flashrom -p ch341a --ifd -i bios -w [path/to/coreboot.rom]
```

### Building

> The building process is the same as with other platforms with coreboot+UEFI.
> Check the [NovaCustom building manual](/unified/novacustom/building-manual/) for
> more detailed instructions.

Reviewer: ```suggestion
Please refer to the [NovaCustom building manual](/unified/novacustom/building-manual/).
```
